### ISSUE TITLE
[question] has anyone successfully configured msvc with ninja and cacche?

### ISSUE THREAD

**Issue by hoyhoy** [conan-io/conan#15452]
### What is your question?

I have it mostly working.  vcvars are loading.  I set my windskver, but something isn't quite right because cl.exe by way of ccache is complaining about PDB file errors immediately.

This is the error I'm seeing...

```
FAILED: src/lib/CMakeFiles/c-ares.dir/ares__htable.c.obj
C:\ccache\cl.exe  /nologo -DCARES_STATICLIB -DHAVE_CONFIG_H=1 -DWIN32_LEAN_AND_MEAN -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_SECURE_NO_DEPRECATE -D_WIN32_WINNT=0x0602 -IC:\Users\tperr\.conan2\p\b\c-are55f61936c5eed\b\build\Debug -IC:\Users\tperr\.conan2\p\b\c-are55f61936c5eed\b\src -IC:\Users\tperr\.conan2\p\b\c-are55f61936c5eed\b\src\include -IC:\Users\tperr\.conan2\p\b\c-are55f61936c5eed\b\src\src\lib /FS /DWIN32 /D_WINDOWS /W3 /W3 /we4013 /Zi /Ob0 /Od /RTC1 -MDd /showIncludes /Fosrc\lib\CMakeFiles\c-ares.dir\ares__htable.c.obj /Fdbin\cares.pdb /FS -c C:\Users\tperr\.conan2\p\b\c-are55f61936c5eed\b\src\src\lib\ares__htable.c
C:\Users\tperr\.conan2\p\b\c-are55f61936c5eed\b\src\src\lib\ares__htable.c(388): fatal error C1090: PDB API call failed, error code '23': (0x00000006)
```

I tried setting various compiler flags like /Zf and the like, but I still keep getting this error.

This is my host profile.

```
[settings]
os=Windows
arch=x86_64
compiler=msvc
compiler.cppstd=14
compiler.version=193
compiler.update=9
compiler.toolset=v141_xp

compiler.cppstd=14
compiler.runtime=dynamic
build_type=Debug

[options]
verbose=True

[conf]
tools.build:compiler_executables = { "c" : "C:\ccache\cl.exe", "cpp": "C:\ccache\cl.exe" }
tools.build:cflags = [ "/FS" ]
tools.build:cxxflags = [ "/FS" ]

tools.microsoft:winsdk_version = 10.0.22621.0
tools.microsoft.msbuild:max_cpu_count = 4
tools.cmake.cmaketoolchain:generator=Ninja
```

Cmake is invoking ccache and ccache is calling the correct cl.exe from msvc, but nothing will build.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by perseoGI** [comment]
---

Hello @hoyhoy,

It has been a while since this issue was opened. Were you able to resolve your problem?

There are some workarounds available for the `ccache` integration issue with Conan. You can follow an interesting discussion on this topic here: https://github.com/conan-io/conan/issues/17754.

Additionally, we have an incubating repository called `conan-toolchains`. I’ve opened a [pull request](https://github.com/conan-io/conan-toolchains/pull/2) to add a new `ccache` recipe that includes a built-in `CMake` toolchain file. This should help configure `ccache` to work properly with the built-in Conan cache.

I encourage you to give it a try and share your feedback. I recently compiled a project using this recipe with `Ninja` as the CMake generator. It was my first time testing this combination on Windows with `ccache`, and it worked well!

```
# Your Windows profile (without any kind of ccache configuration nor anything)
include(default)

[tool_requires]
ccache/4.12

[conf]
# Setup ninja as your CMake generator
tools.cmake.cmaketoolchain:generator=Ninja
```

In the compilation log, using this profile as a host profile, you will get something like:
```
-- ccache found: C:/Users/perseo/.conan2/p/b/ccacha3ec662f1af2b/p/bin/ccache.exe, enabling via CMake launcher and environment.     
-- Using ccache as compiler launcher for Ninja or Makefiles.
```

When using `Ninja` as your generator, you could set the `CMAKE_CXX_COMPILER_LAUNCHER` to `ccache` path, which is what is being done automatically by the above recipe: 

https://github.com/perseoGI/conan-toolchains/blob/602bf2eaecb420ceeeeff3276c9080ab4d195d14/recipes/ccache/binary/ccache-autoinject.cmake?plain=1#L23-L24

I hope this helps!

**Comment by hoyhoy** [comment]
@perseoGI yeah, I could set our package to use ccache, but our dependencies aren't all using cmake.  TBH, I should try adding `CMAKE_CXX_COMPILER_LAUNCHER` to `tools.cmake.cmaketoolchain:extra_variables`.

**Comment by perseoGI** [comment]
Greetings @hoyhoy, 
For the non-CMake scenarios, you need to define `CC/CXX` environment variables to use `ccache`.
In the future, we may attempt to add a native workaround; however, for now, you will need to do it manually.
